crucible/src/Lang/Crucible/CFG/EarlyMergeLoops.hs

```diff
@@ -0,0 +1,839 @@
+--
+-- The module is organized into 3 main components:
+--   1. An analysis that computes the natural loops of a CFG;
+--   2. An analysis that inserts postdominators into loops that have
+--      "early exits" (and hence have postdominators outside the loop);
+--   3. A "fixup" pass that ensures that, in the transformed CFG, all
+--      values are well defined along all (new) paths.
+------------------------------------------------------------------------
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+
+module Lang.Crucible.CFG.EarlyMergeLoops
+  ( earlyMergeLoops
+  ) where
+
+import           Control.Monad ((>=>))
+import           Control.Applicative ((<**>))
+import qualified Data.Graph.Inductive as G
+import qualified Data.Foldable as Fold
+import           Data.Kind
+import           Data.List (nub, minimumBy)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (fromMaybe)
+import           Data.Parameterized.Classes
+import qualified Data.Parameterized.Context as Ctx
+import qualified Data.Parameterized.Map as MapF
+import           Data.Parameterized.Nonce
+import           Data.Parameterized.Some
+import           Data.Parameterized.TraversableFC
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Sequence (Seq)
+import qualified Data.Sequence as Seq ((<|), fromList)
+import           Data.String (fromString)
+
+import           What4.ProgramLoc (Position(..), Posd(..))
+
+import           Lang.Crucible.CFG.Expr
+import           Lang.Crucible.CFG.Reg
+import           Lang.Crucible.Types
+
+--------------------------
+-- | Natural Loop Analysis
+--------------------------
+
+-- | This structure identifies natural loops. Natural loops are either
+-- disjoint from each other, nested, or they share the same header.
+data LoopInfo s = LoopInfo
+  { liFooter     :: !(BlockID s)
+    -- ^ This is the block with a backedge (to the header)
+  , liHeader     :: !(BlockID s)
+    -- ^ This is the destination of the backedge.
+  , liMembers    :: !(Set (BlockID s))
+  -- ^ The loop members, which is the set of nodes that can reach the footer
+  -- without going through the header.
+  , liEarlyExits :: ![CFGEdge s]
+  -- ^ An exiting edge is an edge from a node in the loop to an edge
+  -- not in the loop. An early exit is such an edge from a node that
+  -- is not the footer node.
+  , liFooterIn   :: ![CFGEdge s]
+  -- ^ All the edges to the footer
+  , liDominators :: ![BlockID s]
+  -- ^ The dominators of the loop header.
+  } deriving (Eq, Show)
+type CFGEdge s = (BlockID s, BlockID s)
+
+-- | Detect all loops in a cfg.
+-- The assumption is that two backedges in a cfg will have distinct destination blocks.
```

Comment:
If this assumption is violated, then we return the empty list of loop info and do nothing, is that correct?  Let's state that in this comment.